exporters/etw/include/opentelemetry/exporters/etw/etw_data.h

```diff
@@ -0,0 +1,536 @@
+  /// <returns></returns>
+  virtual void UpdateName(nostd::string_view name) noexcept override
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(name);
+  }
+
+  /// <summary>
+  /// End Span
+  /// </summary>
+  /// <returns></returns>
+  virtual void End(const trace::EndSpanOptions & = {}) noexcept override
+  {
+    // TODO: signal this to owner
+  }
+
+  virtual trace::SpanContext GetContext() const noexcept
+  {
+    // TODO: not implemented
+    static trace::SpanContext nullContext{trace::SpanContext::GetInvalid()};
+    return nullContext;
+  }
+
+  /// <summary>
+  /// Check if Span is recording data
+  /// </summary>
+  /// <returns></returns>
+  virtual bool IsRecording() const noexcept override
+  {
+    // TODO: not implemented
+    return true;
+  }
+
+  virtual void SetToken(nostd::unique_ptr<context::Token> &&token) noexcept
+  {
+    // TODO: not implemented
+    UNREFERENCED_PARAMETER(token);
+  }
+
+  /// <summary>
+  /// Get Owner tracer of this Span
+  /// </summary>
+  /// <returns></returns>
+  trace::Tracer &tracer() const noexcept { return this->owner; };
+};
+
+/// <summary>
+/// stream::TraceProvider
+/// </summary>
+class TracerProvider : public trace::TracerProvider
+{
+public:
+  /// <summary>
+  /// Obtain a Tracer of given type (name) and supply extra argument arg2 to it.
+  /// </summary>
+  /// <param name="name">Tracer Type</param>
+  /// <param name="args">Tracer arguments</param>
+  /// <returns></returns>
+  virtual nostd::shared_ptr<trace::Tracer> GetTracer(nostd::string_view name,
+                                                     nostd::string_view args = "")
+  {
+    // TODO: describe possible args. Currently supported:
+    // "MSGPACK"        - use MessagePack
+    // "XML"            - use XML
+    // "ETW"            - use 'classic' Trace Logging Dynamic manifest ETW event
+    //
+#if defined(HAVE_NO_TLD) && defined(HAVE_MSGPACK)
+    ETWProvider::EventFormat evtFmt = ETWProvider::EventFormat::ETW_MSGPACK;
+#else
+    ETWProvider::EventFormat evtFmt = ETWProvider::EventFormat::ETW_MANIFEST;
+#endif
+
+#pragma warning(push)
+#pragma warning(disable : 4307) /* Integral constant overflow - OK while computing hash */
+    auto h = utils::hashCode(args.data());
+    switch (h)
+    {
+      case CONST_HASHCODE(MSGPACK):
+        // nobrk
+      case CONST_HASHCODE(MsgPack):
+        // nobrk
+      case CONST_HASHCODE(MessagePack):
+        evtFmt = ETWProvider::EventFormat::ETW_MSGPACK;
+        break;
+
+      case CONST_HASHCODE(XML):
+        // nobrk
+      case CONST_HASHCODE(xml):
+        evtFmt = ETWProvider::EventFormat::ETW_XML;
+        break;
+
+      case CONST_HASHCODE(TLD):
+        // nobrk
+      case CONST_HASHCODE(tld):
+        // nobrk
+        evtFmt = ETWProvider::EventFormat::ETW_MANIFEST;
+        break;
+
+      default:
+        break;
+    }
+#pragma warning(pop)
+    return nostd::shared_ptr<trace::Tracer>{new (std::nothrow) Tracer(*this, name, evtFmt)};
```

Comment:
`version` argument as a concept is dubious for statically linked SDK, esp. given in the other thread we are discussing that most likely we are at `v1.0` for very long time; and it will be problematic to provide ABI-stable SDK for `v1.1` , if we need to start supporting new types (e.g. byte arrays). I'd like to keep this as an open item, and actually repurpose the version - making it a mechanism to pass other config params to tracer. Then we do not have to force the tracer provider to support only one format or encoding. It may support different formats / encodings.

Reply:
> We should probably avoid creating a new tracer for each GetTracer call, as this could get expensive...

This is actually optimized as follows:
- Tracer object holds only `encoding` and `provider handle`
- subsequently the `ETWProvider::open` method already maintains the map of registered providers and their handles. And it won't invoke the registration at Win32 API layer, but rather return a wrapper object (`Tracer`) that refers to already registered open handle.

If we are to cache the tracers additionally, not sure we'd save significant CPU time.. Maybe some memory in concurrent use scenario? But I'd argue if users allocate a shared_ptr and forget about it, they should optimize this in their code if that becomes a significant issue. I doubt it's significant, since we already optimize the opening in the `ETWProvider` class.

Reply:
The `version` attribute here doesn't refer to the version of the SDK, but to the version of the instrumentation library (this means, the library that is using OpenTelemetry to add instrumentation), this is the [part of the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#get-a-tracer) describing this.

So, it would look like this:
```c++
auto tracer = provider->GetTracer("io.opentelemetry.contrib.mongodb", "4.4");
```

This would, ideally, be invoked inside MongoDB to obtain a tracer for adding instrumentation.

This would, one day, allow the SDK to enable/disable instrumentation for certain libraries and library versions based on some setting or configuration. If we abuse the `name` or `version` parameter, we'll break this.

Also, your approach will not work. Imagine a third-party library that is instrumented with OpenTelemetry and doesn't know anything about ETW (for example, imagine a `GetTracer` call in nginx). There's no way you can provide the encoding information via the `version` parameter in those instances. The current design here is broken and needs to be fixed.

Reply:
I also just realized that the tracer name is used as ETW provider name. This is broken, for the same reasons given above.